client/rpc/src/cache.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+// This file is part of Frontier.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use codec::Encode;
+use lru::LruCache;
+
+pub struct LRUCacheByteLimited<K, V> {
+	cache: LruCache<K, V>,
+	max_size: u64,
+	metrics: Option<LRUCacheByteLimitedMetrics>,
+	size: u64,
+}
+
+impl<K: Eq + core::hash::Hash, V: Encode> LRUCacheByteLimited<K, V> {
+	pub fn new(
+		cache_name: &'static str,
+		max_size: u64,
+		prometheus_registry: Option<prometheus_endpoint::Registry>,
+	) -> Self {
+		let metrics = match prometheus_registry {
+			Some(registry) => match LRUCacheByteLimitedMetrics::register(cache_name, &registry) {
+				Ok(metrics) => Some(metrics),
+				Err(e) => {
+					log::error!(target: "eth-cache", "Failed to register metrics: {:?}", e);
+					None
+				}
+			},
+			None => None,
+		};
+
+		Self {
+			cache: LruCache::unbounded(),
+			max_size,
+			metrics,
+			size: 0,
+		}
+	}
+	pub fn get(&mut self, k: &K) -> Option<&V> {
+		self.cache.get(k)
+	}
+	pub fn put(&mut self, k: K, v: V) {
+		// Handle size limit
+		self.size += v.encoded_size() as u64;
+		if self.size > self.max_size {
+			let mut diff = self.size - self.max_size;
+			let keys_to_remove = self
+				.cache
+				.iter()
+				.rev()
+				.take_while(|(_k, v)| {
+					let v_size = v.encoded_size() as u64;
+					if diff > v_size {
```

Comment:
Here if the least recently used `v_size` is already greater or equal than the `diff`, no elements will be removed and the new element will be put going over the cache limit.

Instead, in the `take_while`, I'd just return `true` until the cumulative sum of `v_size`s plus `max_size` is greater or equal than `size`:

```
let mut acum = self.max_size;
let keys_to_remove = self
    .cache
    .iter()
    .rev()
    .take_while(|(_k, v)| {
        let v_size = v.encoded_size() as u64;
        if acum < self.size {
            acum += v_size;
            true
        } else {
            false
        }
    })
    .map(|(k, _v)| k)
    .collect::<Vec<_>>();
```